### ISSUE TITLE
parser.pipe() skips examples when batch_size < 4

### ISSUE THREAD

**Issue by RonRademaker** [explosion/spaCy#3344]
## How to reproduce the behaviour
Hi,

When evaluating the space EntityRecognizer I'm running into performance issues. I use ±30 custom entities with about ±70000 examples. When running this in batches on a V100 GPU (batch size of ±16000) one epoch takes about 1 minute and 20 seconds. After that I want to evaluate both my training and test set to gain insight in what's happening. Obviously, this means I need to run all examples through the model, I use the nlp.pipe(samples) structure to do this with a large batch size and no limits on n_threads. Still, this process takes about 8 minutes on the same machine. I think this is very very weird as I don't see how it makes sense that a forward prop with a backprop can be a magnitude faster than just a forward prop. This is why I classify this a but, sorry if it's not and I'm doing something stupid.

Example code:

```
counter = 0
for doc in nlp.pipe(texts, batch_size=16384):
    counter = counter + 1 # Noop to make sure the inside of my loop is not the bottleneck
```

I tried other batch sizes (anything from 2**4 to 2**18) but found no significant difference. Also tried explicit n_threads, but it gives similar results. Using nlp.pipe over looping over example by example did speed up by a factor 2.

## Your Environment

* Operating System: Ubuntu (AWS deep learning AMI, p3.2xlarge)
* Python Version Used: 3.6.5
* spaCy Version Used: 2.0.18
* Environment Information: Tesla V100 GPU with 16GB RAM

**Comment by honnibal** [comment]
Can you try again on `spacy-nightly`? There's some weirdness with the batch sizing in the v2.0 versions. v2.1 should be out soon.

**Comment by RonRademaker** [comment]
Thanks, I tried the nightly version but it looks like entity training broke down somewhere. Testing with the script as documented (I added one sentence that is 100% the same as a training example) on https://spacy.netlify.com/usage/training#ner I get:

```
Created blank 'en' model
Losses {'ner': 29.85830795764923}
Losses {'ner': 22.974769711494446}
Losses {'ner': 15.957626104354858}
Losses {'ner': 9.803914964199066}
Losses {'ner': 8.14788521034643}
Losses {'ner': 8.616836511762813}
Losses {'ner': 7.976333131315187}
Losses {'ner': 7.303826485993341}
Losses {'ner': 4.836598809459247}
Losses {'ner': 3.989627659902908}
Entities in 'Do you like horses?'
Entities in 'they pretend to care about your feelings, those horses'
```

**Comment by RonRademaker** [comment]
By the way, it does look like evaluation is now a matter of seconds 🎉 (but with 0 recall....)

**Comment by syllog1sm** [comment]
The links are pointing to the master branch I think, which will be correct on launch. It should work with the version of the script on develop.

**Comment by RonRademaker** [comment]
Thanks, that script does work for me so obviously I'm doing something wrong. I can't really find any real differences that would explain it though, debugging I did something that strikes me as very odd:

After each epoch I do:
```
print(nlp.get_pipe('ner').model._mem.weights)
```

And it consistently gives me an empty list `[]` which seems to me that the model has no parameters (so it can't predict or learn anything). 

Any ideas where I can look next?

**Comment by syllog1sm** [comment]
Well, the example has very little data, so the learning is pretty delicate. I think the way the batching is different on the master branch, it just happens to not converge well.

As for the weights: try model.upper._mem. The parser model has three layers that hold the weights. See the spacy/syntax/_parser_model.pyx file for details.

**Comment by RonRademaker** [comment]
Thanks, that allowed me to make sure that weights were updating. I also found the issue, it was not with training but with evaluating in batches:

Works as expected:
```
for text in samples:
    doc = nlp(text)
    # evaluate doc.ents
```

Zero recall:
```
for doc in nlp.pipe(texts, batch_size=1):
    # evaluate doc.ents
```

Zero recall:
```
for doc in nlp.pipe(texts, batch_size=2):
    # evaluate doc.ents
```

Works as expected:
```
for doc in nlp.pipe(texts, batch_size=4):
    # evaluate doc.ents
```

Increasing the batch size further things keep working.

**Comment by honnibal** [comment]
Looks suspicious for sure! Thanks for reporting.

**Comment by honnibal** [comment]
Yeah the batching bug was a regression introduced quite recently, when I fixed a constant in the batch sizing. I changed it to take `batch_size // 4` as a value for a sub-batch, but this breaks for `n < 4`. Fixed now.
